File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> pre;
        int res=0,currsum=0;
        for (int i=0;i<nums.size();i++){
            currsum+=nums[i];
            if (currsum==k){
                res++;
            }
            if (pre.find(currsum-k)!=pre.end()){
                res+=pre[currsum-k];
            }
            pre[currsum]++;
            
        }
        return res;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> pre=new HashMap<>();
        pre.put(0,1);
        int res=0,currsum=0;
        for (int i=0;i<nums.length;i++){
            currsum+=nums[i];
            if(pre.containsKey(currsum-k)){
                res+=pre.get(currsum-k);
            }
            pre.put(currsum,pre.getOrDefault(currsum,0)+1);    
        }
        return res;
    }
}